src/Google.Api.Gax/Platform.cs

```diff
@@ -0,0 +1,167 @@
+﻿/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Net.Http;
+using System.Threading;
+using System.Net;
+
+namespace Google.Api.Gax
+{
+    public enum PlatformType
+    {
+        Unknown = 0,
+        Gce = 1,
+        Gae = 2,
+        // TODO: Add GKE if/when possible.
+    }
+
+    public sealed class GcePlatformDetails
+    {
+        public GcePlatformDetails(string metadataJson)
+        {
+            MetadataJson = metadataJson;
+            var json = Newtonsoft.Json.Linq.JObject.Parse(metadataJson);
+            ProjectId = json["project"]["projectId"].ToString();
+            InstanceId = json["instance"]["id"].ToString();
+            ZoneName = json["instance"]["zone"].ToString();
+        }
+
+        public string MetadataJson { get; }
+        public string ProjectId { get; }
+        public string InstanceId { get; }
+        public string ZoneName { get; }
+    }
+
+    public sealed class GaePlatformDetails
+    {
+        public GaePlatformDetails(string gcloudProject, string gaeInstance, string gaeService, string gaeVersion)
+        {
+            ProjectId = gcloudProject;
+            InstanceId = gaeInstance;
+            ServiceId = gaeService;
+            VersionId = gaeVersion;
+        }
+
+        public string ProjectId { get; }
+        public string InstanceId { get; }
+        public string ServiceId { get; }
+        public string VersionId { get; }
+    }
+
+    /// <summary>
+    /// Information about the current execution platform.
+    /// Supported execption platforms are Google App Engine (GAE) and Google Compute Engine (GCE).
+    /// </summary>
+    public sealed class Platform
+    {
+        private readonly static Lazy<Task<Platform>> s_instance = new Lazy<Task<Platform>>(LoadInstance);
+
+        /// <summary>
+        /// Asyncrhonously get execution platform information.
+        /// </summary>
+        /// <returns>A task containing the execution platform information.</returns>
+        public static Task<Platform> InstanceAsync() => s_instance.Value;
+
+        /// <summary>
+        /// Get execution platform information. This may block for up to one second.
+        /// </summary>
+        /// <returns>Execution platform information.</returns>
+        public static Platform Instance() => InstanceAsync().Result;
+
+        private static async Task<GcePlatformDetails> LoadGceDetails()
+        {
+            // TODO: Respect METADATA_EMULATOR_HOST environment variable. Where is this documented?
```

Comment:
I'm not sure I understand. Without checking for the variable, anyone testing locally with the emulator running would get no GCE details. The emulator listens on a localhost port, not on metadata.google.internal. Are you saying that this should intentionally not return info from the emulator?

All that said, the emulator is not a part of the Cloud SDK, so I don't know if this needs to be supported at this time anyway. But without this support, it makes it more difficult for someone to use this (or another) emulator for local testing.